File: challenge/cpp/toh.cpp
#include <iostream>
using namespace std;

int report;

void shuffle(int plates, int from, int to, int use) {

    if (plates == 1) {
        printf("move %d %d\n", from, to);
        return;
    }

    shuffle(plates - 1, from, use, to);
    printf("move %d %d\n", from, to);


    shuffle(plates - 1, use, to, from);
}

int main() {
	int plates;
    printf("Number of plates: ");
    scanf("%d", &plates);
    printf("\n");
    shuffle(plates, 1, 3, 2);
	return 0;
}

File: challenge/go/toh/toh.go
package main

import "fmt"

func shuffle(plates int, from, to, using *[]int) {

	if plates == 1 {
		n := len(*from) - 1
		*to = append(*to, (*from)[n])
		*from = (*from)[:n]
        return
	}

	shuffle(plates-1, from, using, to)
	n := len(*from) - 1
	*to = append(*to, (*from)[n])
	*from = (*from)[:n]

	fmt.Println("From: ", *from)
	fmt.Println("To: ", *to)
	fmt.Println("Using: ", *using)

	shuffle(plates-1, using, to, from)
}

func main() {
	var tests, plates int = 1, 5
	var from, to, using []int

	for ; tests > 0; tests-- {
		for i := plates; i > 0; i-- {
			from = append(from, i)
		}
		shuffle(plates, &from, &to, &using)
	}
}
